components/builder-depot/src/server.rs

```diff
@@ -461,32 +466,76 @@ fn download_latest_origin_encryption_key(req: &mut Request) -> IronResult<Respon
         OriginPublicEncryptionKey,
     >(req, &request) {
         Ok(key) => key,
-        Err(err) => return Ok(render_net_error(&err)),
+        Err(err) => {
+            if err.get_code() == ErrCode::ENTITY_NOT_FOUND {
+                match generate_origin_encryption_keys(&origin, req) {
+                    Ok(key) => key,
+                    Err(Error::NetError(e)) => return Ok(render_net_error(&e)),
+                    Err(_) => unreachable!(),
+                }
+            } else {
+                return Ok(render_net_error(&err));
+            }
+        }
     };
 
     let xfilename = format!("{}-{}.pub", key.get_name(), key.get_revision());
     download_content_as_file(key.get_body(), xfilename)
 }
 
-fn generate_origin_encryption_keys(req: &mut Request) -> IronResult<Response> {
-    debug!("Generate Origin Encryption Keys {:?}", req);
-    match get_param(req, "origin") {
-        Some(origin) => {
-            if !check_origin_access(req, &origin).unwrap_or(false) {
-                return Ok(Response::with(status::Forbidden));
-            }
+fn generate_origin_encryption_keys(
+    origin: &str,
+    req: &mut Request,
+) -> Result<OriginPublicEncryptionKey> {
+    debug!("Generate Origin Encryption Keys {:?} for {:?}", req, origin);
 
-            match helpers::get_origin(req, origin) {
-                Ok(origin) => {
-                    match helpers::generate_origin_encryption_keys(req, origin) {
-                        Ok(_) => Ok(Response::with(status::Created)),
-                        Err(err) => Ok(render_net_error(&err)),
-                    }
-                }
-                Err(err) => Ok(render_net_error(&err)),
+    match helpers::get_origin(req, origin) {
+        Ok(origin) => {
+            let mut public_request = OriginPublicEncryptionKeyCreate::new();
+            let mut private_request = OriginPrivateEncryptionKeyCreate::new();
+            let mut public_key = OriginPublicEncryptionKey::new();
+            let mut private_key = OriginPrivateEncryptionKey::new();
+            {
+                let session = req.extensions.get::<Authenticated>().unwrap();
+                public_key.set_owner_id(session.get_id());
+                private_key.set_owner_id(session.get_id());
             }
+            public_key.set_name(origin.get_name().to_string());
+            public_key.set_origin_id(origin.get_id());
+            private_key.set_name(origin.get_name().to_string());
+            private_key.set_origin_id(origin.get_id());
+
+            let pair = BoxKeyPair::generate_pair_for_origin(origin.get_name())
+                .expect("failed to generate origin encryption key pair");
```

Comment:
Since the function we're in returns a `Result`, is it worth returning an `Err` here instead of calling `expect`?  I remember there was some conversation around the use of `expect` with the crypto functions, and I'm guessing that's how we ended up here.  It looks like the error module in this crate already has an enum for errors from the core crate, so it might work to just turn this into:

```rust
let pair = BoxKeyPair::generate_pair_for_origin(origin.get_name()).map_err(Error::HabitatCore)?;
```

If that doesn't work though, I wouldn't pursue it further.  It looked like an easy win, but maybe it's not.